### PR TITLE
Adding Ablation

### PR DESCRIPTION
**Pull request by ssheikholeslami**

### CODE REVIEW

File: maggy/core/rpc.py [hunk 1]
```diff
@@ -1,7 +1,6 @@
-
 import threading
 import struct
-import pickle
+import cloudpickle
```
Comment (moritzmeister) [inline maggy/core/rpc.py]: Use PySpark's cloudpickle, to avoid having to add cloudpickle additionally to the conda base environment.
```suggestion
from pyspark import cloudpickle
```

File: maggy/experiment.py [hunk 2]
```diff
@@ -72,9 +79,8 @@ def lagom(map_fun, searchspace=None, optimizer=None, direction='max', num_trials
         combination with it's performance metric
     :rtype: dict
     """
-    assert num_trials > 0, "number of trials should be greater than zero"
-
     global running
+    # exp_driver = None  # TODO to fix UnboundLocalError (calling exp_driver.stop() in finally block)
```
Comment (moritzmeister) [inline maggy/experiment.py]: Yes that's something that's on my list. Either remove this line and comment in line 178 `# if running is True and exp_driver is not None:` or propose fix.

File: maggy/experiment.py [hunk 3]
```diff
@@ -29,9 +29,14 @@
...
-    """Launches a maggy experiment for hyperparameter optimization.
-
+def lagom(map_fun, experiment_type,
+          name, hb_interval=1,
+          num_trials=None, searchspace=None, optimizer=None, direction=None,
+          ablation_study=None, ablator=None,
+          es_policy='median', es_interval=300, es_min=10, description=''):
+    """Launches a maggy experiment, which depending on `experiment_type` can
+    either be a hyperparameter optimization or an ablation study experiment.
+
```
Comment (moritzmeister) [inline maggy/experiment.py]: remove space characters
```suggestion

```

File: maggy/experiment.py [hunk 4]
```diff
@@ -104,16 +110,37 @@ def lagom(map_fun, searchspace=None, optimizer=None, direction='max', num_trials
...
+            assert num_trials > 0, "number of trials should be greater than zero"
+            if num_executors > num_trials:
+                num_executors = num_trials
+
+            exp_driver = ExperimentDriver('optimization', searchspace=searchspace, optimizer=optimizer,
+                                          direction=direction, num_trials=num_trials, name=name,
+                                          num_executors=num_executors, hb_interval=hb_interval, es_policy=es_policy,
+                                          es_interval=es_interval, es_min=es_min, description=description,
+                                          app_dir=app_dir, log_dir=log_dir, trial_dir=trial_dir)
```
Comment (moritzmeister) [inline maggy/experiment.py]: One additional indentation level is enough, arguments don't need to align with opening parenthesis of the function call. This also makes the lines too long (>79). But not tragic.
```suggestion
            exp_driver = ExperimentDriver('optimization', searchspace=searchspace,
                optimizer=optimizer, direction=direction, num_trials=num_trials,
                name=name, num_executors=num_executors, hb_interval=hb_interval,
                es_policy=es_policy, es_interval=es_interval, es_min=es_min,
                description=description, app_dir=app_dir, log_dir=log_dir,
                trial_dir=trial_dir)
```

File: maggy/experiment.py [hunk 5]
```diff
@@ -104,16 +110,37 @@ def lagom(map_fun, searchspace=None, optimizer=None, direction='max', num_trials
...
+
+        elif experiment_type == 'ablation':
+            exp_driver = ExperimentDriver('ablation', ablation_study=ablation_study, ablator=ablator,
+                                          name=name, num_executors=num_executors,
+                                          hb_interval=hb_interval, description=description,
+                                          app_dir=app_dir, log_dir=log_dir, trial_dir=trial_dir)
```
Comment (moritzmeister) [inline maggy/experiment.py]: Same as above.

File: maggy/experiment.py [hunk 6]
```diff
@@ -128,8 +155,8 @@ def lagom(map_fun, searchspace=None, optimizer=None, direction='max', num_trials
 
         # Force execution on executor, since GPU is located on executor
         job_start = datetime.now()
-        nodeRDD.foreachPartition(trialexecutor._prepare_func(app_id, run_id,
-            map_fun, server_addr, hb_interval, exp_driver._secret, app_dir))
+        nodeRDD.foreachPartition(trialexecutor._prepare_func(app_id, run_id, experiment_type,
+                                 map_fun, server_addr, hb_interval, exp_driver._secret, app_dir))
```
Comment (moritzmeister) [inline maggy/experiment.py:158]: Same as above.

File: maggy/experiment.py [hunk 7]
```diff
@@ -148,13 +175,15 @@ def lagom(map_fun, searchspace=None, optimizer=None, direction='max', num_trials
         # sparkmagic hb poll intervall is 5 seconds, therefore wait 6 seconds
         time.sleep(6)
         # cleanup spark jobs
+        # if running is True and exp_driver is not None:
```
Comment (moritzmeister) [inline maggy/experiment.py]: Remove or propose change.

File: maggy/ablation/ablator/loco.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
+from maggy.ablation.ablator import AbstractAblator
+from maggy.ablation.ablationstudy import AblationStudy
+from hops import featurestore
+from hops import hdfs as hopshdfs
+import tensorflow as tf
+from maggy.trial import Trial
+import json
+
+
+class LOCO(AbstractAblator):
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py:11]: All capital names are usually used for constants.
Class names should normally follow the CapWords convention.
But it's up to you.

File: maggy/core/experimentdriver.py [hunk 9]
```diff
@@ -7,13 +7,17 @@
...
 from datetime import datetime
+from sys import maxsize as integer_max
 from maggy import util
 from maggy.optimizer import AbstractOptimizer, RandomSearch, Asha, SingleRun
 from maggy.core import rpc
 from maggy.trial import Trial
 from maggy.earlystop import AbstractEarlyStop, MedianStoppingRule, NoStoppingRule
 from maggy.searchspace import Searchspace
 
+from maggy.ablation.ablator import AbstractAblator, LOCO
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py:18]: Class names should normally use the CapWords convention.

File: maggy/core/experimentdriver.py [hunk 10]
```diff
@@ -24,98 +28,185 @@
 class ExperimentDriver(object):
 
     SECRET_BYTES = 8
+    # TODO rewrite with @classmethod
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Let's make this a todo on the Trello board. So we can remove the comment.

File: maggy/core/experimentdriver.py [hunk 11]
```diff
@@ -24,98 +28,185 @@
...
+        self._trial_store = {}
+        self.num_executors = kwargs.get('num_executors')
+        self._message_q = queue.Queue()
+        self.name = kwargs.get('name')
+        self.experiment_done = False
+        self.worker_done = False
+        self.hb_interval = kwargs.get('hb_interval')
+        self.description = kwargs.get('description')
+        self.experiment_type = experiment_type
+        self.es_interval = integer_max  # XXX not the cleanest way
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Could we also set default `es_policy` to `None` in `lagom()`?
Or in general not sure we need it and you can just set it both to the default, if user doesn't provide custom values because of line 150.
```suggestion
        self.es_interval = kwargs.get(es_interval)		
        self.es_min = kwargs.get(es_min)
```

You could also set it to None in case of ablation and add an if statement before line 296.

File: maggy/core/experimentdriver.py [hunk 12]
```diff
@@ -24,98 +28,185 @@
...
+            if isinstance(direction, str) and direction.lower() in ['min', 'max']:
+                self.direction = direction.lower()
+            else:
+                raise Exception(
+                    "The experiment's direction should be an string (either 'min' or 'max') "
+                    "but it is {0} (of type '{1}')."
+                    .format(str(direction), type(direction).__name__))
+
+            es_policy = kwargs.get('es_policy')
+            # XXX check self.early_stop_check vs. self.earlystop_check
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: I think we talked about that. We never used `self.early_stop_check`.

File: maggy/core/experimentdriver.py [hunk 13]
```diff
@@ -24,98 +28,185 @@
...
+                self.ablator = ablator
+                print("Custom Ablator initialized. \n")
+            else:
+                raise Exception(
+                    "The experiment's ablation study policy should either be a string ('loco') "
+                    "or a custom policy that is an instance of maggy.ablation.ablation.AbstractAblator, "
+                    "but it is {0} (of type '{1}')."
+                    .format(str(ablator), type(ablator).__name__))
+
+            # XXX setup ablation result schema
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Remove XXX, or do you need to change something more?

File: maggy/core/experimentdriver.py [hunk 14]
```diff
@@ -124,32 +215,53 @@ def init(self):
...
+                'Total job time ' + self.duration + '\n'
 
-        self.duration = hopsutil._time_diff(self.job_start, self.job_end)
+        elif self.experiment_type == 'ablation':
 
+            _ = self.ablator.finalize_experiment(self._final_store)
+            self.job_end = datetime.now()
+            self.duration = hopsutil._time_diff(self.job_start, self.job_end)
+
+            # TODO modify so that "None" is not printed
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Remove or fix :)

File: maggy/core/experimentdriver.py [hunk 15]
```diff
@@ -124,32 +215,53 @@ def init(self):
...
-            'AVERAGE metric -- ' + str(self.result['avg']) + '\n' \
-            'EARLY STOPPED Trials -- ' + str(self.result['early_stopped']) + '\n' \
-            'Total job time ' + self.duration + '\n'
         print(results)
 
         self._log(results)
 
         hopshdfs.dump(json.dumps(self.result, default=util.json_default_numpy),
-            self.app_dir + '/result.json')
+                      self.app_dir + '/result.json')
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py:265]: One level of indentation is enough. Change back.

File: maggy/core/experimentdriver.py [hunk 16]
```diff
@@ -244,10 +355,15 @@ def _target_function(self):
                         self.maggy_log = self._update_maggy_log()
                         self._log(self.maggy_log)
 
-                        hopshdfs.dump(trial.to_json(), self.trial_dir + '/' + trial.trial_id + '/trial.json')
+                        if self.experiment_type == 'optimization':
+                            hopshdfs.dump(trial.to_json(), self.trial_dir + '/' + trial.trial_id + '/trial.json')
+                        # XXX what about ablation? forgot to dump to json?
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Fix or remove.

File: maggy/core/experimentdriver.py [hunk 17]
```diff
@@ -344,38 +471,67 @@ def _update_result(self, trial):
...
+            param_string.pop('dataset_function', None)
+            param_string.pop('model_function', None)
+            # First finalized trial
+            if self.result.get('best_id', None) is None:
+                self.result = {
+                    'best_id': trial_id, 'best_val': metric, 'best_config': param_string,
+                    'worst_id': trial_id, 'worst_val': metric, 'worst_config': param_string,
+                    'avg': metric, 'metric_list': [metric], 'num_trials': 1,
+                    'early_stopped': 0,
+                    # TODO earlystop included temporarily for compatibility with sparkmagic, remove before release
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Not sure what needs to be done here.

File: maggy/core/experimentdriver.py [hunk 18]
```diff
@@ -344,38 +471,67 @@ def _update_result(self, trial):
...
+                }
+                return
+
+            # for ablation we always consider 'max' as "the direction"
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py:523]: What if user uses loss?

File: maggy/core/experimentdriver.py [hunk 19]
```diff
@@ -388,12 +544,21 @@ def _update_maggy_log(self):
...
+
+        if self.experiment_type == 'optimization':
+            log = 'Maggy Optimization ' + str(finished) + '/' + str(self.num_trials) + \
+                ' (' + str(self.result['early_stopped']) + ') ' + \
+                util._progress_bar(finished, self.num_trials) + ' - BEST ' + \
+                json.dumps(self.result['best_hp']) + ' - metric ' + \
+                str(self.result['best_val'])
+
+        elif self.experiment_type == 'ablation':
+            # TODO modify so that "None" is not printed
```
Comment (moritzmeister) [inline maggy/core/experimentdriver.py]: Remove or fix.

File: maggy/ablation/ablator/loco.py [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+            pass  # TODO process and change the custom dataset generator
+        else:
+            training_dataset_name = self.ablation_study.hops_training_dataset_name
+            training_dataset_version = self.ablation_study.hops_training_dataset_version
+            label_name = self.ablation_study.label_name
+
+            if dataset_type == 'tfrecord':
+
+                def create_tf_dataset(num_epochs, batch_size):
+                    # TODO @Moritz: go with shadowing? i.e., def create_tf_dataset(ablated_feature)?
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py]: I think we talked about it, but don't remember what we decided.

File: maggy/ablation/ablator/loco.py [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+                    SHUFFLE_BUFFER_SIZE = 10000  # XXX parametrize?
```
Comment (moritzmeister) [inline maggy/ablation/ablator/loco.py]: Define on module level?

File: maggy/ablation/ablationstudy.py [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+            else:
+                self._include_single_feature(arg)
+
+    def _include_single_feature(self, feature):
+        if type(feature) is str:
+            self.included_features.add(feature)  # TODO should check with the list retrieved from the featurestore
+        else:
+            raise ValueError("features.include() only accepts strings or lists of strings, "
+                             "but it received {0} which is of type '{1}'."
+                             .format(str(feature), type(feature).__name__))
```
Comment (moritzmeister) [inline maggy/ablation/ablationstudy.py]: Maybe change indenting:
```suggestion
            raise ValueError(
                "features.include() only accepts strings or lists of strings, "
                "but it received {0} which is of type '{1}'."
                .format(str(feature), type(feature).__name__))
```

**Review by moritzmeister — CHANGES_REQUESTED** [review]
I mainly reviewed the changed in the core and experiment modules.
Tested the changes with the MNIST notebook, only change required is the addition of the `experiment_type` argument when launching `lagom()`.
The only functional change really necessary is the switch to PySpark's cloudpickle.
The other comments are mostly about indentation.

Other than that I would just go through all the TODOs and XXX comments, to see if they can be fixed, removed or added to Trello as a Todo.

File: maggy/core/experimentdriver.py (hunk 18 above)
Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py:523]: For now let's keep it that way, since eventually this is only used for showing the trial with `best_val` in the notebook, and the assumption is that the users are interested in accuracy. But the point totally makes sense, and they might want to use e.g. loss so `best_val` should correspond to the lowest loss.

Anyhow, I will explain this in the documentation and will add this as a TODO to the board, so we can come up with a better way of handling this.

File: maggy/core/experimentdriver.py (hunk 11 above)
Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py]: I added `if self.earlystop_check == NoStoppingRule.earlystop_check: ` before `if (datetime.now() - time_earlystop_check).total_seconds() >= self.es_interval: `. This should work since for ablation I already set its `earlystop_check` to `NoStoppingRule.earlystop_check`, but you should double check this again.

Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py]: https://github.com/logicalclocks/maggy/pull/27/commits/2b6977cebde9eb67bbd09db082a31904951f2591#diff-bfdfef7fbe9209260b7ba10581f534edR289

**Comment by ssheikholeslami** [comment]
For now I set 'optimization' as the default 'experiment_type' so as not to break existing notebooks and documentation.

**Comment by ssheikholeslami** [comment]
LGTM, @moritzmeister and @Limmen please go through the changes once more.

File: maggy/ablation/ablationstudy.py [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+                for feature in arg:
+                    self._exclude_single_feature(feature)
+            else:
+                self._exclude_single_feature(arg)
+
+    def _exclude_single_feature(self, feature):
+        if type(feature) is str:
+            if feature in self.included_features:
+                self.included_features.remove(feature)
+                print("Feature '{0}' is excluded from the ablation study.")
```
Comment (Kim-Hammar) [inline maggy/ablation/ablationstudy.py]: is not `.format(str(feature))` missing here?

File: maggy/ablation/ablator/loco.py [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+        if self.ablation_study.custom_dataset_generator:
+            pass
+        else:
+            training_dataset_name = self.ablation_study.hops_training_dataset_name
+            training_dataset_version = self.ablation_study.hops_training_dataset_version
+            label_name = self.ablation_study.label_name
+
+            if dataset_type == 'tfrecord':
+
+                def create_tf_dataset(num_epochs, batch_size):
```
Comment (Kim-Hammar) [inline maggy/ablation/ablator/loco.py:34]: Will this work if the dataset_type is not tfrecord? Maybe throw an "NotSupportedException" if the user pass another dataset type

**Review by Kim-Hammar — APPROVED** [review]

**Comment by Kim-Hammar** [comment]
LGTM @ssheikholeslami @moritzmeister, minor review points only.

File: maggy/ablation/ablationstudy.py (hunk 23 above)
Comment (ssheikholeslami) [reply maggy/ablation/ablationstudy.py]: indeed! :D

Comment (ssheikholeslami) [reply maggy/ablation/ablationstudy.py]: Fixed with
https://github.com/logicalclocks/maggy/pull/27/commits/05e4efa4894526ed2e275b34929d4593bbad6cb9

**Review by moritzmeister — APPROVED** [review]
All good!

File: maggy/ablation/ablator/loco.py (hunk 24 above)
Comment (ssheikholeslami) [reply maggy/ablation/ablator/loco.py:34]: In such cases the users have to provide their own dataset generator functions (for now). Added a custom exception, and it will be raised there for now.
https://github.com/logicalclocks/maggy/pull/27/commits/6f2ed1a2dd3465c4ad54974a6e1e32a6ab797be7

Comment (Kim-Hammar) [reply maggy/ablation/ablator/loco.py:34]: Ok, good!
